aiida_mlip/data/model.py

```diff
@@ -0,0 +1,224 @@
+"""Define Model Data type in AiiDA."""
+
+import hashlib
+from pathlib import Path
+from typing import Any, Optional, Union
+from urllib import request
+from urllib.parse import urlparse
+
+from aiida.orm import SinglefileData
+
+
+class ModelData(SinglefileData):
+    """Class to save a model file as an AiiDA data type.
+
+    The file can be a file that is stored locally or a new file to download.
+    """
+
+    @staticmethod
+    def _calculate_hash(file: Union[str, Path]) -> str:
+        """Calculate the hash of a file.
+
+        Parameters
+        ----------
+        file : Union[str, Path]
+            Path to the file for which hash needs to be calculated.
+
+        Returns
+        -------
+        str
+            The SHA-256 hash of the file.
+        """
+        # Calculate hash
+        buf_size = 65536  # reading 64kB (arbitrary) at a time
+        sha256 = hashlib.sha256()
+        with open(file, "rb") as f:
+            # calculating sha in chunks rather than 1 large pass
+            while data := f.read(buf_size):
+                sha256.update(data)
+        file_hash = sha256.hexdigest()
+        return file_hash
+
+    @staticmethod
+    def _check_existing_file(
+        file: Union[str, Path],
+    ) -> Path:
+        """Check if a file already exists and return the path of the existing file if it does.
+
+        Parameters
+        ----------
+        file : Union[str, Path]
+            Path to the downloaded model file.
+
+        Returns
+        -------
+        Path
+            The path of the model file of interest (same as input path if no duplicates were found).
+        """
+        file_hash = ModelData._calculate_hash(file)
+
+        def is_diff_file(curr_path: Path):
+            return curr_path.is_file() and not curr_path.samefile(file)
+
+        file_folder = Path(file).parent
+        for existing_file in filter(is_diff_file, file_folder.rglob("*")):
+            if ModelData._calculate_hash(existing_file) == file_hash:
+                ex_file_folder = Path(existing_file).parent
+                if file_folder == ex_file_folder:
+                    file.unlink()
+                    return Path(existing_file)
+        return Path(file)
```

Comment:
I may be misreading, but am I correct in thinking that if you download `/cache/mace/file`, this will search for existing files matching `file` in  `/cache/` recursively?

Then, if the file already exists in `/cache/mace/`, it discards the new file, but are we not ignoring matches in any other directory (say, `/cache/mace_mp/file`). That would be picked up as `existing_file` (I think?), but that information is unused?

Maybe we should throw a warning or something (if we still want to check all other architecture directories, which doesn't seem unreasonable)?

Reply:
So, initially, the idea was that if you have `architecture` it will only check in `cache_dir/architecture` (so if there is the same model in `cache_dir/` it will not catch it), while if you don't have `architecture` it will check in all the subfolders of `cache_dir`, so if it finds a duplicate in a subfolder, it automatically knows what architecture it is (and this is the only reason that recursive check made sense).
Now that I made `architecture` mandatory, I would expect a folder `cache_dir` to not have any files but just subfolders with the names of different architectures. So basically, the check will happen inside the single subfolder, not in `cache_dir`.

So yes, if there are duplicates outside the folder that has been checked, we will miss them, but in theory there should not be (cause I am assuming that a specific model only works for a specific architecture, which I'm not 100% sure it's true, and that every file in the cache_dir has an architecture already assigned, so it's already in the right folder, and that everything in cache_dir is created with the ModelData class, which is also not necessarily true, cause one could  add things to it manually... but in that case that's their fault?)

Reply:
Do we still need `if file_folder == ex_file_folder` then if we're only checking the same directory?

I don't think we necessarily need to worry about what people could do manually, but I can imagine slightly different names for the same model, but maybe that's a problem for the user.